### ISSUE TITLE
"Play to" provides no way to order files

### ISSUE THREAD

**Issue by mcarlton00** [jellyfin/jellyfin#1724]
**Describe the bug**
When using the 'Play to' feature, there's currently no way for the receiving client to know what order files should be played in.

**To Reproduce**
1. Be logged into 2 clients that support the 'Play to' (web client, android app, kodi)
2. Send a playlist/collection from one device to another
4. Playlist/collection contents play in a seemingly random order

**Expected behavior**
Items should play in the same order as displayed

**Logs**
Mixed logs and findings from investigating
https://pastebin.com/8zct2g76

**System (please complete the following information):**
 - OS: Docker
 - Browser: Firefox
 - Jellyfin Version: 10.3.7
 - Reverse proxy: nginx

**Additional context**
Originally discovered while investigating https://github.com/jellyfin/jellyfin-kodi/issues/49, but this also occurs with the mobile client or 2 different web clients.  It seems like originally this was handled by the server sorting item IDs before sending instead of by the clients, but was removed in https://github.com/jellyfin/jellyfin/pull/1011.  Currently, the server only sends a list of ItemIds of the contents of what is being sent (Playlist, Collection, etc).  There's no way for the clients to know what order these items should be played in without more information about the parent.  It seems like either the server needs to sort items before sending them, or it needs to expose the ItemID of what's being sent (Playlist, Collection, etc) so the clients can look up that data themselves.

**Comment by JustAMan** [comment]
ping @Bond-009 as IIRC this is your change that broke sorting... any idea how to fix this situation?

**Comment by Bond-009** [comment]
The items are returned ordered by their `SortName`, since I don't know how the order of a playlist is determined I don't know how to fix this (without reintroducing the issues #1011 fixed)

**Comment by JustAMan** [comment]
Can we enhance the result with item ids? So that a client can do the sorting which was previously done in server?

**Comment by dkanada** [comment]
Maybe if we don't order the return value at all it will show up in the same order the original request started with.

**Comment by JustAMan** [comment]
We don't order them, but their order is DB-dependant (so you can see it as undefined or random, whatever you like more). Old code ordered them by some rules...

**Comment by Germwalker** [comment]
#1011 removed the server side sorting of items IDs to fix a performance issue, but I wonder how sorting less than 20 elements can take so much time & cpu cycle. Maybe we can reimplement sorting in a better way ?

**Comment by dkanada** [comment]
As far as I recall, it was actually one of the largest performance improvements made for the latest release.

**Comment by mcarlton00** [comment]
The same logic applies to playlists, so it could be potentially sorting hundreds of items.  But yeah, still seems odd that it was such a big performance hit.

Would it be possible to just send a list of item IDs in the same order it's displayed in the Web UI, along with an index to let the client know where in that list they should start?  Maybe something like this?
```
{
`payload': {
  'itemIds': ['item1', 'item2', 'item3'],
  'startAt: 1
  }
}
```

**Comment by joshuaboniface** [comment]
We should probably look at reverting #1011, it may have made a slight speedup, but the effects are wide-reaching.

**Comment by Germwalker** [comment]
I agree, this problem is really anoying. Items must be ordered before beeing sent to kodi, if the code that sort items is inefficient it needs to be optimised. But this problem make the jelly / kodi combo completely unusable for non advanced users. The only workaround now is to send the entire album (music) / season (tv show) and sort manualy in the kodi web interface. That's a pain in the ass.

**Comment by Bond-009** [comment]
The speedup is huge “10sec to 230ms” (for rpi3 b+) is VERY noticeable. I don’t think we should revert it.

**Comment by dkanada** [comment]
Maybe a configuration option? I forget how large of a change it was, but this could be viable if it's only a single place.

**Comment by mcarlton00** [comment]
If a sort of 10-20 items for an album or season is taking 10 seconds, it sounds like the problem is the implementation method used, not that this is actually a feature.  Plus the server has all this info and is the source of truth.  It seems as though it should be providing that info, rather than leaving the clients to figure it out on their own

**Comment by dlahoti** [comment]
The real issue seems to have been that the sorting was done using `OrderBy(i => ids.IndexOf(i.Id))`, which takes O(n^2 log n) time, when we have enough information to do it in O(n) time. If the sorting was added back with a better implementation, it probably wouldn't be an issue even on a Raspberry Pi.

**Comment by JustAMan** [comment]
I don't think we can do it in `O(n)`, AFAIK the minimum complexity (proven mathematically) for a (random valued) sorting is `O(n*log n)`, though it is still an order of magnitude better than what is done here.

I hope I can try playing with that later on and see if I can get it.

**Comment by dlahoti** [comment]
That's true for comparison sorting, but we can do this without any comparisons: one pass over `ids` can build a dictionary with the IDs as keys and the positions as values, and then one pass over the unsorted query result can place its items into a new array according to the positions given by the dictionary. Something like
```
var pos = new Dictionary<ItemId, Queue<int>>();
for (int i = 0; i < ids.Count; i++) {
    if (!pos.ContainsKey(ids[i])) {
        pos.Add(ids[i], new Queue<int>());
    } 
    pos[ids[i]].Enqueue(i);
}
var ret = new BaseItem[ids.Count];
foreach (var item in result) {
    ret[pos[item.Id].Dequeue()] = item;
}
```
(The `Queue` is there to make sure things work even when the same item appears multiple times.)

**Comment by JustAMan** [comment]
That isn't completely true, as you're assuming that `pos[item.Id]` does it by O(1). Typical hashmap has the worst lookup complexity as O(n), so your "worst case" complexity here is O(n*n) :)

That said, average lookup is definitely O(1), so _on average_ you're going to have an O(n) result here.

So while I disagree from mathematical standpoint, in observable practice you're probably correct :)
Thanks for suggesting this idea, I'm going to play with it.

**Comment by JustAMan** [comment]
I've made a few benchmarks, and using `Queue` in there as you suggested kills all the performance. My guess is it uses a linked list underneath.

I'm pretty sure I can do away with a `List<>` plus offset for dequeing, but still it doesn't seem worth the hassle - I've tried to make a super dumbed benchmark, and I cannot believe that sorting itself should take more than 20-30 ms for 100 items on Raspberry Pi 3. So it should be something else.
